third_party/nvidia/backend/compiler.py

```diff
@@ -189,8 +189,8 @@ def make_ttir(mod, metadata, opt):
         pm.enable_debug()
         passes.common.add_inliner(pm)
         passes.ttir.add_rewrite_tensor_pointer(pm)
-        passes.ttir.add_combine(pm)
         passes.common.add_canonicalizer(pm)
+        passes.ttir.add_combine(pm)
```

Comment:
I think that makes sense. I wonder if this will prevent some canonicalization to happen as broadcast ops may be in the middle. What kind of canonicalization do you need for this pass?
Worst case if we see problems we can run canonicalization one more time

Reply:
I think all the broadcast related things are now in canonicalize and combine is now mainly dot-related things, plus the `addptr(addptr(ptr, a), b) -> addptr(ptr, a + b)` pattern which shouldn't be effected by broadcasting really. So I think this is a good separation.

I do wonder if it might be worthwhile to add another canonicalize pass though, since LICM and loop unrollng could potentially connect patterns that were previously separated by region boundaries. The reorder broadcast pass could also connect arith patterns that were broken up by broadcasts and/or splats.

Just a hunch though, I don't have any examples in mind.